### ISSUE TITLE
Thread deadlock problem in agent's destroy function

### ISSUE THREAD

**Issue by gallandarakhneorg** [sarl/sarl#458]
_From @ngaud on November 14, 2014 16:14_

Big problem of thread deadlock when you call a skill in destroy function.

Typical use case:

```
behavior MyBehavior {

    on XX  {  
        killMe
    }
}

agent BehaviorPlayer {

    uses Behaviors

    var myBehavior : MyBehavior


    on Initialize {

        myBehavior = new MyBehavior(this)
        this.registerBehavior(myFactorialBehavior);
        ...
    } 

    on Destroy {
        this.unregisterBehavior(myBehavior);
    }


}

```

killMe -> killAgent -> InternalEventBusSkill$AgentListener.killOwner -> iterate over skills uninstallation -> uninstall ExternalContextAccessSkill -> fire context left -> requires BehaviorSkill.wake (synchronized)

In parallel killMe fires destroy event, destroy is executed that calls BehaviorSkill.unregisterBehavior calling InternalEventBusSkill.unregisterEventListener

Basically one thread owns InternalEventBusSkill and waits for BehaviorSkill, and the other owns BehaviorSkill and waits for InternalEventBusSkill: typical deadlock.

The problem is the parallelization of tasks with destroy function.

_Copied from original issue: janus-project/janusproject#89_

**Comment by gallandarakhneorg** [comment]
This issue may be partly solved by the issue #98.

**Comment by gallandarakhneorg** [comment]
This issue will be moved into the SARL project according to sarl/sarl#447.
